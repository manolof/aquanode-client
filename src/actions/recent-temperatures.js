import { connect } from 'socket.io-client';

const socket = connect(process.env.REACT_APP_ENDPOINT_RECENT_TEMPERATURES);

export const RecentTemperaturesActions = Object.freeze({
	GET: '[RecentTemperatures] GET',
});

export const getRecentTemperaturesAction = () => (dispatch) => {
	socket.on('get', (message) => {
		dispatch({
			type: RecentTemperaturesActions.GET,
			payload: message.data,
		});
	});
};
