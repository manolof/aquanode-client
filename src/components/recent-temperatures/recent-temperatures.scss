.temperatures-graph-wrapper {
	position: relative;
}
