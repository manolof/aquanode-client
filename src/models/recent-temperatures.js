import PropTypes from 'prop-types';

const recentTemperatureReading = PropTypes.shape({
	date: PropTypes.string,
	temperature: PropTypes.number,
});

export const RecentTemperaturesModel = PropTypes.shape({
	recentTemperatures: PropTypes.arrayOf(recentTemperatureReading),
});
