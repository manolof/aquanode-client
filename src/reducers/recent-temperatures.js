import { RecentTemperaturesActions } from '../actions/recent-temperatures';

export const initialState = {
	readings: [],
};

const recentTemperaturesReducer = (state = initialState, action) => {
	switch (action.type) {
		case RecentTemperaturesActions.GET:
			return {
				readings: action.payload,
			};

		default:
			return state;
	}
};

export {
	recentTemperaturesReducer,
};
