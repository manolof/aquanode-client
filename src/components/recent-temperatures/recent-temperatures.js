import React, { Component } from 'react';
import moment from 'moment-timezone';
import { Line } from 'react-chartjs-2';

import './recent-temperatures.scss';
import { RecentTemperaturesModel } from '../../models/recent-temperatures';

function factorData(data) {
	return data.map((e, i, a) => {
		const prev = a[i - 1];
		const next = a[i + 1];
		if (e === prev && e === next) {
			return '' + e;
		}
		return e;
	}).map(e => typeof e === 'string' ? null : e);
}

class RecentTemperatures extends Component {
	formatDateTime(dateTime) {
		return moment.tz(dateTime, 'Europe/Zurich').format('MMM DD, HH:mm');
	}

	render() {
		const { recentTemperatures } = this.props;
		const { readings } = recentTemperatures;

		const labels = readings.reverse().map((sample) => this.formatDateTime(sample.date));
		const datasetData = readings.reverse().map((dataset) => dataset.temperature);

		const datasets = [
			{
				label: 'Temperature (°C)',
				data: factorData(datasetData),
				fill: false,
				lineTension: 0.1,
				borderColor: 'rgb(255, 99, 132)',
				spanGaps: true,
			},
		];
		const data = {
			labels,
			datasets,
		};

		const options = {
			responsive: true,
			legend: {
				display: false,
			},
			tooltips: {
				mode: 'index',
				intersect: false,
			},
			hover: {
				mode: 'nearest',
				intersect: true,
			},
			scales: {
				xAxes: [{
					display: true,
					scaleLabel: {
						display: false,
					},
				}],
				yAxes: [{
					display: true,
					scaleLabel: {
						display: true,
						labelString: 'Temperature (°C)',
					},
					ticks: {
						suggestedMin: 24,
						stepSize: .5,
					},
				}],
			},
		};

		return (
			<div className="panel">
				<header className="panel__header">
					<h3>Recent Temperatures</h3>
				</header>

				<main className="panel__content">
					<section className="temperatures-graph-wrapper">
						<Line
							data={data}
							height={100}
							options={options}
						/>
					</section>
				</main>
			</div>
		);
	}
}

RecentTemperatures.propTypes = {
	recentTemperatures: RecentTemperaturesModel.isRequired,
};

export {
	RecentTemperatures,
};
