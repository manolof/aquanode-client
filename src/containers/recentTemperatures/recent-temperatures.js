import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import { getRecentTemperaturesAction } from '../../actions/recent-temperatures';
import { RecentTemperatures } from '../../components/recent-temperatures/recent-temperatures';
import { RecentTemperaturesModel } from '../../models/recent-temperatures';

class RecentTemperaturesContainer extends Component {
	componentDidMount() {
		const { getRecentTemperatures } = this.props;
		getRecentTemperatures();
	}

	render() {
		const { recentTemperatures } = this.props;

		return (
			<RecentTemperatures recentTemperatures={recentTemperatures}/>
		);
	}
}

RecentTemperaturesContainer.propTypes = {
	recentTemperatures: RecentTemperaturesModel.isRequired,
	getRecentTemperatures: PropTypes.func.isRequired,
};

// istanbul ignore next
const mapStateToProps = (state) => ({
	recentTemperatures: state.recentTemperatures,
});

// istanbul ignore next
const mapDispatchToProps = (dispatch) => ({
	getRecentTemperatures: () => dispatch(getRecentTemperaturesAction()),
});

const RecentTemperaturesContainerConnect = connect(mapStateToProps, mapDispatchToProps)(RecentTemperaturesContainer);

export {
	RecentTemperaturesContainer,
	RecentTemperaturesContainerConnect,
};
