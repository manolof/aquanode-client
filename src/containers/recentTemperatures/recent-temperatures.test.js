import { shallow } from 'enzyme';
import React from 'react';

import { RecentTemperaturesContainer } from './recent-temperatures';

describe('RecentTemperatures Container', () => {
	let wrapper;
	const getRecentTemperaturesSpy = jest.fn();
	const mockRecentTemperaturesProps = {
		readings: [
			{
				date: '2018-10-23',
				temperature: 42,
			},
		],
	};

	beforeEach(() => {
		wrapper = shallow(
			<RecentTemperaturesContainer
				recentTemperatures={mockRecentTemperaturesProps}
				getRecentTemperatures={getRecentTemperaturesSpy}
			/>,
		);
	});

	afterEach(() => {
		jest.resetAllMocks();
	});

	afterAll(() => {
		jest.restoreAllMocks();
	});

	it('should render', () => {
		expect(wrapper.getElements()).toMatchSnapshot();
	});

	it('should get the recent temperatures on load', () => {
		expect(getRecentTemperaturesSpy).toHaveBeenCalled();
	});

});
