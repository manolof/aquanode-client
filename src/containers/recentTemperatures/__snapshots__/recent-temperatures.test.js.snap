// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecentTemperatures Container should render 1`] = `
Array [
  <RecentTemperatures
    recentTemperatures={
      Object {
        "readings": Array [
          Object {
            "date": "2018-10-23",
            "temperature": 42,
          },
        ],
      }
    }
  />,
]
`;
